crates/chia-sdk-driver/src/layers/p2_eip712_message_layer.rs

```diff
@@ -0,0 +1,298 @@
+        self.construct_spend(
+            ctx,
+            P2Eip712MessageSolution {
+                my_id,
+                signed_hash: self.hash_to_sign(my_id, ctx.tree_hash(delegated_spend.puzzle).into()),
+                signature: Bytes::new(signature.to_vec()),
+                delegated_puzzle: delegated_spend.puzzle,
+                delegated_solution: delegated_spend.solution,
+            },
+        )
+    }
+
+    pub fn domain_separator(&self) -> Bytes32 {
+        let type_hash = keccak256(b"EIP712Domain(string name,string version,bytes32 salt)");
+
+        keccak256(ethers::abi::encode(&[
+            ethers::abi::Token::FixedBytes(type_hash.to_vec()),
+            ethers::abi::Token::FixedBytes(keccak256("Chia Coin Spend").to_vec()),
+            ethers::abi::Token::FixedBytes(keccak256("1").to_vec()),
+            ethers::abi::Token::FixedBytes(self.genesis_challenge.to_vec()),
+        ]))
+        .into()
+    }
+
+    pub fn prefix_and_domain_separator(&self) -> [u8; 34] {
+        let mut pads = [0u8; 34];
+        pads[0] = 0x19;
+        pads[1] = 0x01;
+        pads[2..].copy_from_slice(&self.domain_separator());
+        pads
+    }
+
+    pub fn type_hash() -> Bytes32 {
+        keccak256(b"ChiaCoinSpend(bytes32 coin_id,bytes32 delegated_puzzle_hash)").into()
+    }
+
+    pub fn hash_to_sign(&self, coin_id: Bytes32, delegated_puzzle_hash: Bytes32) -> Bytes32 {
+        /*
+        bytes32 messageHash = keccak256(abi.encode(
+            typeHash,
+            coin_id,
+            delegated_puzzle_hash
+        ));
+        */
+        let mut to_hash = Vec::new();
+        to_hash.extend_from_slice(&P2Eip712MessageLayer::type_hash());
+        to_hash.extend_from_slice(&coin_id);
+        to_hash.extend_from_slice(&delegated_puzzle_hash);
+
+        let message_hash = keccak256(&to_hash);
+
+        let mut to_hash = Vec::new();
+        to_hash.extend_from_slice(&self.prefix_and_domain_separator());
+        to_hash.extend_from_slice(&message_hash);
+
+        keccak256(&to_hash).into()
+    }
+}
+
+impl Layer for P2Eip712MessageLayer {
+    type Solution = P2Eip712MessageSolution<NodePtr, NodePtr>;
+
+    fn construct_puzzle(&self, ctx: &mut SpendContext) -> Result<NodePtr, DriverError> {
+        let curried = CurriedProgram {
+            program: ctx.p2_eip712_message_puzzle()?,
+            args: P2Eip712MessageArgs {
+                prefix_and_domain_separator: self.prefix_and_domain_separator().to_vec().into(),
+                type_hash: P2Eip712MessageLayer::type_hash(),
+                pubkey: self.pubkey.to_vec().into(),
+            },
+        };
+        ctx.alloc(&curried)
+    }
+
+    fn construct_solution(
+        &self,
+        ctx: &mut SpendContext,
+        solution: Self::Solution,
+    ) -> Result<NodePtr, DriverError> {
+        ctx.alloc(&solution)
+    }
+
+    fn parse_puzzle(_allocator: &Allocator, _puzzle: Puzzle) -> Result<Option<Self>, DriverError> {
+        Ok(None)
```

Comment:
Is it not possible to parse this?

Reply:
no - can't recover genesis challenge since it's hashed into the `PREFIX_AND_DOMAIN_SEPARATOR`; do you think we could bruteforce it? :) (i.e., try the challenges for mainnet/testnet11; return none if no hashes match up)

Reply:
updated as we talked on Discord & the parse method works now :)